File: ios/MetanavCore/Sources/MetanavCore/Labels.swift
import Foundation

/// Friendly names and typical heights for detector labels (COCO vocabulary).
public enum Labels {
    public static let generic = "Obstacle"

    private static let friendly: [String: String] = [
        "person": "Person", "bicycle": "Bicycle", "car": "Car", "motorcycle": "Motorbike", "motorbike": "Motorbike",
        "bus": "Bus", "train": "Train", "truck": "Truck", "boat": "Boat", "traffic light": "Traffic light",
        "fire hydrant": "Fire hydrant", "stop sign": "Sign", "parking meter": "Parking meter", "bench": "Bench",
        "dog": "Dog", "cat": "Cat", "horse": "Horse", "chair": "Chair", "couch": "Couch", "sofa": "Couch",
        "potted plant": "Plant", "pottedplant": "Plant", "bed": "Bed", "dining table": "Table", "diningtable": "Table",
        "toilet": "Toilet", "tv": "TV", "tvmonitor": "TV", "refrigerator": "Fridge", "suitcase": "Suitcase",
        "backpack": "Bag", "handbag": "Bag", "umbrella": "Umbrella", "bottle": "Bottle", "vase": "Vase",
        "sink": "Sink", "oven": "Oven", "microwave": "Microwave", "laptop": "Laptop", "skateboard": "Skateboard",
        "sports ball": "Ball",
    ]

    /// Approximate real-world heights in meters used for size-based distance estimates.
    private static let heights: [String: Float] = [
        "person": 1.70, "bicycle": 1.00, "motorcycle": 1.10, "motorbike": 1.10, "car": 1.50, "bus": 3.20,
        "truck": 3.00, "chair": 0.90, "bench": 0.85, "fire hydrant": 0.75, "dog": 0.55, "potted plant": 0.70,
        "pottedplant": 0.70, "dining table": 0.75, "diningtable": 0.75, "couch": 0.85, "sofa": 0.85,
        "refrigerator": 1.75, "suitcase": 0.65, "traffic light": 1.00, "stop sign": 0.75,
    ]

    public static func friendlyName(_ raw: String) -> String {
        let key = raw.trimmingCharacters(in: .whitespaces).lowercased()
        if let name = friendly[key] { return name }
        if key.isEmpty { return generic }
        return key.prefix(1).uppercased() + key.dropFirst()
    }

    public static func knownHeightMeters(_ raw: String) -> Float? {
        heights[raw.trimmingCharacters(in: .whitespaces).lowercased()]
    }
}

File: android/core/src/main/kotlin/com/metanav/core/Labels.kt
package com.metanav.core

/** Friendly names and typical heights for detector labels (COCO vocabulary). */
object Labels {
    private val friendly = mapOf(
        "person" to "Person",
        "bicycle" to "Bicycle",
        "car" to "Car",
        "motorcycle" to "Motorbike",
        "motorbike" to "Motorbike",
        "bus" to "Bus",
        "train" to "Train",
        "truck" to "Truck",
        "boat" to "Boat",
        "traffic light" to "Traffic light",
        "fire hydrant" to "Fire hydrant",
        "stop sign" to "Sign",
        "parking meter" to "Parking meter",
        "bench" to "Bench",
        "dog" to "Dog",
        "cat" to "Cat",
        "horse" to "Horse",
        "chair" to "Chair",
        "couch" to "Couch",
        "sofa" to "Couch",
        "potted plant" to "Plant",
        "pottedplant" to "Plant",
        "bed" to "Bed",
        "dining table" to "Table",
        "diningtable" to "Table",
        "toilet" to "Toilet",
        "tv" to "TV",
        "tvmonitor" to "TV",
        "refrigerator" to "Fridge",
        "suitcase" to "Suitcase",
        "backpack" to "Bag",
        "handbag" to "Bag",
        "umbrella" to "Umbrella",
        "bottle" to "Bottle",
        "vase" to "Vase",
        "sink" to "Sink",
        "oven" to "Oven",
        "microwave" to "Microwave",
        "laptop" to "Laptop",
        "skateboard" to "Skateboard",
        "sports ball" to "Ball",
    )

    /** Approximate real-world heights in meters used for size-based distance estimates. */
    private val heights = mapOf(
        "person" to 1.70f,
        "bicycle" to 1.00f,
        "motorcycle" to 1.10f,
        "motorbike" to 1.10f,
        "car" to 1.50f,
        "bus" to 3.20f,
        "truck" to 3.00f,
        "chair" to 0.90f,
        "bench" to 0.85f,
        "fire hydrant" to 0.75f,
        "dog" to 0.55f,
        "potted plant" to 0.70f,
        "pottedplant" to 0.70f,
        "dining table" to 0.75f,
        "diningtable" to 0.75f,
        "couch" to 0.85f,
        "sofa" to 0.85f,
        "refrigerator" to 1.75f,
        "suitcase" to 0.65f,
        "traffic light" to 1.00f,
        "stop sign" to 0.75f,
    )

    const val GENERIC = "Obstacle"

    fun friendlyName(raw: String): String {
        val key = raw.trim().lowercase()
        return friendly[key] ?: key.replaceFirstChar { it.uppercase() }.ifBlank { GENERIC }
    }

    fun knownHeightMeters(raw: String): Float? = heights[raw.trim().lowercase()]
}
